Vpn modules

lukas-verm opened the pull request:
Add modules to send out VPN requests including:

- TCP module: Establish TCP connection and send out probe
- SSTP module: Establish TLS connection and send out SSTP_DUPLEX_POST HTTP request

lukas-verm commented on modules/sstp/scanner.go:
```diff
@@ -0,0 +1,228 @@
+	Status         string `json:"status,omitempty"`
+	Content_Length string `json:"content_length,omitempty"`
+	Host           string `json:"host,omitempty"`
+	Content_Type   string `json:"content_type,omitempty"`
```
To analyze SSTP response, we actually need only the status, content length and host, so it would be possible to remove the rest. This would make the output much cleaner, however, our raw data also includes those headers. The modules would differ to when we used them for our analysis then.

ogasser replied:
I think it's fine to keep it as it is :+1:

lukas-verm commented on modules/tcp/scanner.go:
```diff
@@ -0,0 +1,234 @@
+
+				//generate session ID
+				//the ID is static in order to identify and verify incoming responses
+				session := "1a2b3c4d1a2b3c4d"
```
It makes sense to use a static session ID in order to identify and verify the responses, but it seems a little weird to do that when publishing the modules. Should we maybe give the user to set a custom session ID?

ogasser replied:
You can implement this as a config parameter and have this session ID specified as the default there.

ogasser commented on modules/sstp/scanner.go:
```diff
@@ -0,0 +1,228 @@
+
+	err = conn.Handshake()
+
+	if conn != nil {
```
In what scenarios would `conn` be nil? Since two methods have already been called for it, I don't think this will ever happen, would it?

If we need this, this should probably be moved up right after `conn := tls.Client(tcp_conn, cfg)`.

ogasser commented on modules/sstp/scanner.go:
```diff
@@ -0,0 +1,228 @@
+	for i, elem := range splits {
+		if i == 0 { //first element is the status message
+			status = elem
+			continue
```
Isn't it better to simply use `if`, `else if`, … `else` instead of all `if`s and then adding a `continue`? With the first option we don't need the `continue`, right?

ogasser commented on modules/sstp/scanner.go:
```diff
@@ -0,0 +1,228 @@
+	return zgrab2.SCAN_SUCCESS, &results, nil
+}
+
+// RegisterModule is called by modules/tcp.go to register this module with the
```
Replace tcp.go with sstp.go

ogasser commented on modules/tcp/scanner.go:
```diff
@@ -0,0 +1,234 @@
+	zgrab2.TLSFlags
+
+	Protocol  string `long:"protocol" default:"\\n" description:"Send an initiation request of the VPN protocol specified here." `
+	ProbeFile string `long:"probe-file" description:"Read probe from file as byte array (hex). Mutually exclusive with --probe"`
```
Shouldn't this be `--protocol`?

ogasser commented on modules/tcp/scanner.go:
```diff
@@ -0,0 +1,234 @@
+// Validate performs any needed validation on the arguments
+func (flags *Flags) Validate(args []string) error {
+	if flags.Protocol != "\\n" && flags.ProbeFile != "" {
+		log.Fatal("Cannot set both --probe and --probe-file")
```
Shouldn't this be `--protocol`?

ogasser commented on modules/tcp/scanner.go:
```diff
@@ -0,0 +1,234 @@
+	zgrab2.BaseFlags
+	zgrab2.TLSFlags
+
+	Protocol  string `long:"protocol" default:"\\n" description:"Send an initiation request of the VPN protocol specified here." `
```
What are possible options for Protocol?

lukas-verm replied:
This can be either "pptp" or "openvpn". I added some information on the session parameters and on how to use the module in the README.md.

ogasser replied:
I suggest to put these two possible values in the description here as well.

ogasser replied:
Put the two possible values in the description of the `Protocol` flag as well.

lukas-verm replied:
I did that in the latest version, the commit you are referencing here is outdated.

ogasser replied:
Right, then it looks good to me :+1:

ogasser commented on modules/tcp/scanner.go:
```diff
@@ -0,0 +1,234 @@
+					opcode = "08"
+				}
+
+				if scanner.config.KeyMethod == 2 {
```
Shouldn't this be `else if` for better readability?

ogasser commented on modules/tcp/scanner.go:
```diff
@@ -0,0 +1,234 @@
+					length = "000e"
+					strProbe = length + opcode + session + "0000000000"
+				}
+				if f.HMAC { //hmac generation
```
This should probably also be `else if`, right?

ogasser commented on modules/tcp/scanner.go:
```diff
@@ -0,0 +1,234 @@
+	)
+
+	//try to open connection for specified amount of tries
+	for try < scanner.config.MaxTries {
```
I would write this as follows:

```
for try := 0; try < scanner.config.MaxTries; try++ {
```

ogasser commented on modules/tcp/scanner.go:
```diff
@@ -0,0 +1,234 @@
+	//send probe to open connection and check for response
+	var ret []byte
+	try = 0
+	for try < scanner.config.MaxTries {
```
I would write this as follows:

```
for try := 0; try < scanner.config.MaxTries; try++ {
```

ogasser reviewed (changes requested):
Code looks good! I've added a couple of minor comments.

maganiss commented on README.md:
```diff
@@ -152,5 +152,20 @@ $ echo $?
+ * `--keymethod`: Specifies OpenVPN key method. This can be eith 1 or 2.
+ * `--protocol`: Specifies which VPN protocol to scan for. Allows us to omit a probe. This can be either *`openvpn`* or *`pptp`*.
+ * `--hmac`: Specify this flag to include HMAC, packet ID and timestamp in an OpenVPN request.
+ * `--probefile`: File which contains a custom probe to be sent over the TCP connection.
```
It makes sense to add a sample probefile to the repo, either here in the readme same as what you have for `multiple.ini` , or in just a separate file and a reference here in readme.

lukas-verm replied:
Since the probefile does not require a particular format and only contains the probe, I added a small example in the README.md VPN section instead of adding an example file.

maganiss replied:
Looks good.

maganiss commented on modules/sstp/scanner.go:
```diff
@@ -0,0 +1,215 @@
+	}
+
+	//send probe and read response
+	probe := getProbe(host)
```
I assume this will override the --probefile option if selected. If so, it makes sense to note this in the readme and/or in the comment above this line. Same thing applies to PPTP.

lukas-verm replied:
For SSTP, you cannot specify a probe file. For PPTP, you can use the protocol option or a probe file, they are mutually exclusive. You will receive a warning when trying to use both and this is mentioned now for both flags when opening the TCP help options.

lukas-verm commented:
I addressed your comments and made a minor change to the probefile. By default when reading out a file, the probe had to be specified in an unintuitive unicode sequence. I changed this to except hex streams, which also allows to easily export probes from, e.g., Wireshark which can then be passed in a probefile.

I also realized that the TCP module included TLS flags which were not used. I removed them so the help mask for the TCP module looks clean and only gives information on the flags that can actually be used.

ogasser reviewed (approved):
Looks good to me :+1:
